### PR TITLE
GC data on S3

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6827

Problem Summary: run GC on S3 data to reclaim the storage usage

### What is changed and how it works?

Depends on
* https://github.com/pingcap/tiflash/pull/6878
* https://github.com/pingcap/tiflash/pull/6863
* https://github.com/tikv/client-c/pull/126
* https://github.com/pingcap/tiflash/pull/6874
* https://github.com/pingcap/tiflash/pull/6900

Components

* CheckpointManifestS3Set - parse the manifests list info from S3
* S3GCManager. For each store_id on S3
  * Get the latest manifest from S3
  * Parse the locks hold by the latest manifest (depend on Manifest uploading)
  * Scan and remove the expired locks, try add delmark when the data file is not locked by any store (depend on S3LockService)
  * Remove the data file with expired delmark

TODOs:
- [ ] Remove the sub files for DTFile
- [ ] Collect valid locks from multiple manifest
- [ ] Write a mark to skip useless `LIST` for tombstone store files
- [ ] Calculate the valid size of CheckpointDataFile and compact them
- [ ] Use DataStoreS3 for downloading manifest file

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/Universal/UniversalWriteBatchImpl.h [hunk 1]
```diff
@@ -103,20 +106,56 @@ class UniversalWriteBatch : private boost::noncopyable
     void putExternal(const UniversalPageId & page_id, UInt64 tag)
     {
         // External page's data is not managed by PageStorage, which means data is empty.
-        Write w{WriteBatchWriteType::PUT_EXTERNAL, page_id, tag, nullptr, 0, "", {}};
+        Write w{WriteBatchWriteType::PUT_EXTERNAL, page_id, tag, nullptr, 0, "", {}, std::nullopt};
         writes.emplace_back(std::move(w));
     }
 
+    void putRemotePage(PageIdU64 page_id, const PS::V3::CheckpointLocation & loc, PageSize size, const PageFieldSizes & data_sizes = {})
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/Universal/UniversalWriteBatchImpl.h:113]: I can't find the place this method is used.

File: dbms/src/Storages/S3/S3GCManager.cpp [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+        locks.merge(part_locks.value());
+    }
+
+    return locks;
+}
+
+void S3GCManager::removeOutdatedManifest(const CheckpointManifestS3Set & manifests, const Aws::Utils::DateTime * const timepoint)
+{
+    // clean the outdated manifest files
+    auto expired_bound_sec = config.manifest_expired_hour * 3600;
```
Comment (flowbehappy) [inline dbms/src/Storages/S3/S3GCManager.cpp:420]: How about also retaining the last 10 manifests?

File: dbms/src/Storages/S3/CheckpointManifestS3Set.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+namespace DB::S3
+{
+CheckpointManifestS3Set
+CheckpointManifestS3Set::getFromS3(const S3::TiFlashS3Client & client, StoreID store_id)
+{
+    const auto store_prefix = S3::S3Filename::fromStoreId(store_id).toManifestPrefix();
+
+    std::vector<CheckpointManifestS3Object> manifests;
+
+    listPrefix(client, client.bucket(), store_prefix, [&](const Aws::S3::Model::ListObjectsV2Result & result) {
```
Comment (flowbehappy) [inline dbms/src/Storages/S3/CheckpointManifestS3Set.cpp:28]: We need some metrics about how many times S3 API has been called.

Comment (flowbehappy) [reply dbms/src/Storages/S3/CheckpointManifestS3Set.cpp:28]: Need to wait for this PR https://github.com/pingcap/tiflash/pull/6913/files#diff-a9609b7aa8c2b0399aeccb4cd74379bceaea166374dd71247e81f7ba50671a36R286

File: dbms/src/Storages/Page/V3/Universal/UniversalWriteBatchImpl.h (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/Universal/UniversalWriteBatchImpl.h]: removed

File: dbms/src/Storages/S3/S3GCManager.cpp (hunk 2 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3GCManager.cpp]: updated

File: dbms/src/Storages/S3/tests/gtest_s3gcmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        auto mock_pd_client = std::make_shared<pingcap::pd::MockPDClient>();
+        gc_mgr = std::make_unique<S3GCManager>(mock_pd_client, mock_s3_client, mock_gc_owner, mock_lock_client, config);
+    }
+
+    std::shared_ptr<tests::MockS3Client> mock_s3_client;
+    std::unique_ptr<S3GCManager> gc_mgr;
+    LoggerPtr log;
+};
+
+TEST_F(S3GCManagerTest, RemoveManifest)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/tests/gtest_s3gcmanager.cpp:64]: TODO: the ut need to adapt to current mock s3 client

File: dbms/src/Storages/S3/CheckpointManifestS3Set.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            // outdated manifest objects
+            manifests.emplace_back(CheckpointManifestS3Object{mf_key, object.GetLastModified()});
+        }
+        return DB::S3::PageResult{.num_keys = objects.size(), .more = true};
+    });
+    return CheckpointManifestS3Set::create(std::move(manifests));
+}
+
+CheckpointManifestS3Set
+CheckpointManifestS3Set::create(std::vector<CheckpointManifestS3Object> manifest_keys)
```
Comment (lidezhu) [inline dbms/src/Storages/S3/CheckpointManifestS3Set.cpp:44]:
```suggestion
CheckpointManifestS3Set::create(std::vector<CheckpointManifestS3Object> & manifest_keys)
```

File: dbms/src/Storages/S3/CheckpointManifestS3Set.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    String key;
+    Aws::Utils::DateTime last_modification;
+};
+
+class CheckpointManifestS3Set
+{
+public:
+    static CheckpointManifestS3Set getFromS3(const S3::TiFlashS3Client & client, StoreID store_id);
+
+    static CheckpointManifestS3Set create(std::vector<CheckpointManifestS3Object> manifest_keys);
```
Comment (lidezhu) [inline dbms/src/Storages/S3/CheckpointManifestS3Set.h:41]:
```suggestion
    static CheckpointManifestS3Set create(std::vector<CheckpointManifestS3Object> & manifest_keys);
```

File: dbms/src/Storages/S3/S3GCManager.h [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    void physicalRemoveDataFile(const String & datafile_key);
+
+    std::vector<UInt64> getAllStoreIds() const;
+
+    std::unordered_set<String> getValidLocksFromManifest(const String & manifest_key);
+
+    void removeOutdatedManifest(const CheckpointManifestS3Set & manifests, const Aws::Utils::DateTime * const timepoint); // NOLINT(readability-avoid-const-params-in-decls)
+
+    String getTemporaryDownloadFile(String s3_key);
```
Comment (lidezhu) [inline dbms/src/Storages/S3/S3GCManager.h:129]:
```suggestion
    String getTemporaryDownloadFile(const String & s3_key);
```

File: dbms/src/Storages/S3/S3GCManager.cpp [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+        deleteObject(*client, client->bucket(), mf.key);
+        LOG_INFO(
+            log,
+            "remove outdated manifest, key={} mtime={}",
+            mf.key,
+            mf.last_modification.ToGmtString(Aws::Utils::DateFormat::ISO_8601));
+    }
+}
+
+String S3GCManager::getTemporaryDownloadFile(String s3_key)
```
Comment (lidezhu) [inline dbms/src/Storages/S3/S3GCManager.cpp:454]:
```suggestion
String S3GCManager::getTemporaryDownloadFile(const String & s3_key)
```

File: dbms/src/Storages/S3/S3GCManager.cpp [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+    //       in the next round.
+
+    // After all the locks removed, the data files may still being locked by another
+    // store id, we need to scan the data files with expired delmark
+    tryCleanExpiredDataFiles(gc_store_id, gc_timepoint);
+}
+
+void S3GCManager::cleanUnusedLocks(
+    UInt64 gc_store_id,
+    String scan_prefix,
```
Comment (lidezhu) [inline dbms/src/Storages/S3/S3GCManager.cpp:184]:
```suggestion
    const String & scan_prefix,
```

File: dbms/src/Storages/S3/S3GCManager.h [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    void shutdown() { shutdown_called = true; }
+
+    // private:
+    void runForStore(UInt64 gc_store_id);
+
+    void runForTombstonedStore(UInt64 gc_store_id);
+
+    void cleanUnusedLocks(
+        UInt64 gc_store_id,
+        String scan_prefix,
```
Comment (lidezhu) [inline dbms/src/Storages/S3/S3GCManager.h:106]:
```suggestion
        const String & scan_prefix,
```

**Review by lidezhu — APPROVED** [review]
LGTM with minor comments

File: dbms/src/Storages/S3/S3GCManager.cpp (hunk 8 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3GCManager.cpp:454]: this can not be `const &` because I replace '/' with '_' in the s3_key for a download name

**Pull request merged**